design-documents/asynchronous-operations/consumer-processes-improvements.md

```diff
@@ -0,0 +1,86 @@
+
+There is also the potential problem that the consumer process will take up more and more memory due to some memory leak in the code it executes.
+
+### The suggested solution
+
+The suggestion is to give more control per consumer to the shopowner or developers managing the shop.  
+We could add some additional options to the consumer processes to keep them more under control. Currently there is one available limit available: `max-messages`. If that number of messages was processed, the consumer will kill itself.  
+I'd like to suggest some other limits which we can set:
+
+- `max-idle-time`: if no message was being handled in xx seconds, then kill yourself
+- `max-time`: after xx total seconds, and after you are done with handling the current message, kill yourself
+- `memory-limit`: if xx MB's of memory is being taken up by the process and after you are done with handling the current message, kill yourself (a way to work around potential memory leaks)
+
+Next to these limits, a configurable sleep time might be nice:
+
+- `sleep`: xx milliseconds to sleep before checking if a new message is available (currently this is [hardcoded to 1 second](https://github.com/magento/magento2/blob/2.3.2/lib/internal/Magento/Framework/MessageQueue/CallbackInvoker.php#L59))
+
+I'd also like to see an option defined on the consumer, but being used by the `consumers_runner` cronjob:
+
+- `only-spawn-when-message-available`: the idea is that the `consumers_runner` job checks the queue before spawning a consumer, to see if there is actually a message pending in the queue. If there is one, then go ahead and spawn a consumer (only if one isn't already running). If there isn't a message in the queue, then don't spawn a consumer.
+
+### Some options combined
+
+The problem outlined above, where a specific consumer only needs to run very infrequently could be solved by combining the options:
+
+- `only-spawn-when-message-available`
+- `max-idle-time`
+
+The consumer will only spawn when it is needed, and it will kill itself when it wasn't active for a certain period.  
+That should save some precious server resources.
+
+### Making these options configurable and have some defaults
+
+These options should be configurable per consumer type.
+Some sensible defaults could be set in the [`queue_consumer.xml`](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/message-queues/config-mq.html#queueconsumerxml) file for some of these options.  
+Next to that, developers or shopowners should be able to override these values per consumer type. At least being able to override it in the `app/etc/env.php` file would be nice, but a backend interface for making these things configurable would also be very nice (but can maybe be done in a later phase?).
```

Comment:
Agree with this point. Will be nice to have UI for available consumers, their shedule and even about amount of processed messages